README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

## The `RcppCensSpatial R` package

<!-- badges: start -->
<!-- badges: end -->

The `RcppCensSpatial R` package provides functions to estimate the
parameters in spatial models with censored and missing responses via EM
(Dempster, Laird, and Rubin 1977), SAEM (Delyon, Lavielle, and Moulines
1999), and MCEM (Wei and Tanner 1990) algorithm. These algorithms are
widely used to compute the maximum likelihood (ML) estimates in
incomplete data problems. The EM algorithm computes the ML estimates
when a closed expression for the conditional expectation of the
complete-data log-likelihood function is available. In the MCEM
algorithm, the conditional expectation is substituted by a Monte Carlo
(MC) approximation based on many independent simulations of the missing
data, while the SAEM algorithm splits the E-step into a simulation step
and an integration step. The SAEM algorithm was developed as an
alternative to the computationally intensive MCEM algorithm.

This package approximates the standard error of the estimates using the
method developed by (Louis 1982). It also has a function that performs
spatial prediction in a set of new locations. Besides the functions to
estimate parameters, this package allows calculating the covariance
matrix and the distance matrix.

### Functions

`RcppCensSpatial` package provides the following functions:

-   `CovMat`: Computes the spatial covariance matrix.
-   `dist2Dmatrix`: Computes the Euclidean distance matrix for a set of
    coordinates.
-   `EM.sclm`: Returns the ML estimates of the unknown parameters
    computed via the EM algorithm.
-   `MCEM.sclm`: Returns the ML estimates of the unknown parameters
    computed via the MCEM algorithm.
-   `SAEM.sclm`: Returns the ML estimates of the unknown parameters
    computed via the SAEM algorithm.
-   `predict.sclm`: Performs spatial prediction in a set of new
    locations.
-   `rCensSp`: Simulates censored spatial data for an established
    censoring rate.

Functions `print`, `summary`, `predict`, and `plot` also work for the
`sclm` class.

### Installation

You can install the released version of `RcppCensSpatial` from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("RcppCensSpatial")
```

### References

<div id="refs" class="references csl-bib-body hanging-indent">

<div id="ref-delyon1999convergence" class="csl-entry">

Delyon, Bernard, Marc Lavielle, and Eric Moulines. 1999. “Convergence of
a Stochastic Approximation Version of the EM Algorithm.” *Annals of
Statistics* 27 (1): 94–128. <https://www.jstor.org/stable/120120>.

</div>

<div id="ref-dempster1977maximum" class="csl-entry">

Dempster, Arthur P, Nan M Laird, and Donald B Rubin. 1977. “Maximum
Likelihood from Incomplete Data via the EM Algorithm.” *Journal of the
Royal Statistical Society: Series B (Methodological)* 39 (1): 1–22.
<https://www.jstor.org/stable/2984875>.

</div>

<div id="ref-louis1982finding" class="csl-entry">

Louis, Thomas. 1982. “Finding the Observed Information Matrix When Using
the EM Algorithm.” *Journal of the Royal Statistical Society: Series B
(Methodological)* 44 (2): 226–33.
<https://www.jstor.org/stable/2345828>.

</div>

<div id="ref-wei1990monte" class="csl-entry">

Wei, Greg CG, and Martin A Tanner. 1990. “A Monte Carlo Implementation
of the EM Algorithm and the Poor Man’s Data Augmentation Algorithms.”
*Journal of the American Statistical Association* 85 (411): 699–704.
<https://doi.org/10.1080/01621459.1990.10474930>.

</div>

</div>


NEWS.md
# RcppCensSpatial 0.1.0

* Initial release.


R/Algorithm_utils.R

# Estimate parameters using MCEM algorithm
# -----------------------------------------------------------------------------
MCEM_Spatial = function(y,x,cens,LI,LS,coords,init.phi,init.nugget,type.sc,kappa,
                        lower,upper,MaxIter,nMin,nMax,tol,show.SE){
  if (sum(cens)==0) {
    t1 = Sys.time()
    output = Spatial_model(y,x,coords,init.phi,init.nugget,lower,upper,type.sc,kappa,MaxIter,tol,show.SE)
    t2 = Sys.time()
    ptime = difftime(t2,t1,units="m")

  } else {
    t1 = Sys.time()
    output = MCEMspatial(y,x,cens,LI,LS,coords,init.phi,init.nugget,lower,upper,type.sc,kappa,
                        MaxIter,nMin,nMax,tol,show.SE)
    t2 = Sys.time()
    ptime = difftime(t2,t1,units="m")
  }
  likeli = log_likelihood(y,cens,LI,LS,x,output$beta,output$sigma2,output$phi,output$tau2,
                              coords,type.sc,kappa)
  if (show.SE){
    output = list(Theta=output$Theta, theta=output$theta, beta=output$beta, sigma2=output$sigma2,
                  phi=output$phi, tau2=output$tau2, EY=output$EY, EYY=output$EYY, SE=output$SE,
                  InfMat=output$InfMat, loglik=likeli$loglik, AIC=likeli$AIC, BIC=likeli$BIC,
                  Iterations=output$Iterations, ptime=ptime, X=x, coord=coords, type=type.sc,
                  kappa=kappa)
  } else {
    output = list(Theta=output$Theta, theta=output$theta, beta=output$beta, sigma2=output$sigma2,
                  phi=output$phi, tau2=output$tau2, EY=output$EY, EYY=output$EYY, loglik=likeli$loglik,
                  AIC=likeli$AIC, BIC=likeli$BIC, Iterations=output$Iterations, ptime=ptime, X=x,
                  coord=coords, type=type.sc, kappa=kappa)
  }
  return (output)
}


# Estimate parameters using SAEM algorithm
# -----------------------------------------------------------------------------
SAEM_Spatial = function(y,x,cens,LI,LS,coords,init.phi,init.nugget,type.sc,kappa,
                        lower,upper,MaxIter,M,pc,tol,show.SE){
  if (sum(cens)==0) {
    t1 = Sys.time()
    output = Spatial_model(y,x,coords,init.phi,init.nugget,lower,upper,type.sc,kappa,MaxIter,tol,show.SE)
    t2 = Sys.time()
    ptime = difftime(t2,t1,units="m")

  } else {
    t1 = Sys.time()
    output = SAEMspatial(y,x,cens,LI,LS,coords,init.phi,init.nugget,lower,upper,type.sc,
                         kappa,MaxIter,pc,M,tol,show.SE)
    t2 = Sys.time()
    ptime = difftime(t2,t1,units="m")
  }
  likeli = log_likelihood(y,cens,LI,LS,x,output$beta,output$sigma2,output$phi,output$tau2,
                          coords,type.sc,kappa)
  if (show.SE){
    output = list(Theta=output$Theta, theta=output$theta, beta=output$beta, sigma2=output$sigma2,
                  phi=output$phi, tau2=output$tau2, EY=output$EY, EYY=output$EYY, SE=output$SE,
                  InfMat=output$InfMat, loglik=likeli$loglik, AIC=likeli$AIC, BIC=likeli$BIC,
                  Iterations=output$Iterations, ptime=ptime, X=x, coord=coords, type=type.sc,
                  kappa=kappa)
  } else {
    output = list(Theta=output$Theta, theta=output$theta, beta=output$beta, sigma2=output$sigma2,
                  phi=output$phi, tau2=output$tau2, EY=output$EY, EYY=output$EYY, loglik=likeli$loglik,
                  AIC=likeli$AIC, BIC=likeli$BIC, Iterations=output$Iterations, ptime=ptime, X=x,
                  coord=coords, type=type.sc, kappa=kappa)
  }
  return (output)
}


# Estimate parameters using EM algorithm
# -----------------------------------------------------------------------------
EM_Spatial = function(y,x,cens,LI,LS,coords,init.phi,init.nugget,type.sc,kappa,
                      lower,upper,MaxIter,tol,show.SE){
  if (sum(cens)==0) {
    t1 = Sys.time()
    output = Spatial_model(y,x,coords,init.phi,init.nugget,lower,upper,type.sc,kappa,MaxIter,tol,show.SE)
    t2 = Sys.time()
    ptime = difftime(t2,t1,units="m")

  } else {
    t1 = Sys.time()
    output = EMspatial(y,x,cens,LI,LS,coords,init.phi,init.nugget,lower,upper,type.sc,kappa,MaxIter,tol,show.SE)
    t2 = Sys.time()
    ptime = difftime(t2,t1,units="m")
  }
  likeli = log_likelihood(y,cens,LI,LS,x,output$beta,output$sigma2,output$phi,output$tau2,
                          coords,type.sc,kappa)
  if (show.SE){
    output = list(Theta=output$Theta, theta=output$theta, beta=output$beta, sigma2=output$sigma2,
                  phi=output$phi, tau2=output$tau2, EY=output$EY, EYY=output$EYY, SE=output$SE,
                  InfMat=output$InfMat, loglik=likeli$loglik, AIC=likeli$AIC, BIC=likeli$BIC,
                  Iterations=output$Iterations, ptime=ptime, X=x, coord=coords, type=type.sc,
                  kappa=kappa)
  } else {
    output = list(Theta=output$Theta, theta=output$theta, beta=output$beta, sigma2=output$sigma2,
                  phi=output$phi, tau2=output$tau2, EY=output$EY, EYY=output$EYY, loglik=likeli$loglik,
                  AIC=likeli$AIC, BIC=likeli$BIC, Iterations=output$Iterations, ptime=ptime, X=x,
                  coord=coords, type=type.sc, kappa=kappa)
  }
  return (output)
}


# Log-likelihood function
# -----------------------------------------------------------------------------
log_likelihood = function(y,cc,lower,upper,x,beta,sigma2,phi,tau2,coords,type,kappa){

  mean = x%*%beta
  distM = crossdist(coords)
  Variance = varianceMat(phi,tau2,sigma2,kappa,distM,type)$Sigma

  if (sum(cc)==0) {
    logver = dmvnorm(c(y),c(mean),Variance,TRUE,TRUE)
  } else {
    invObs = solve(Variance[cc==0,cc==0])
    meanC = mean[cc==1] + Variance[cc==1,cc==0]%*%invObs%*%(y[cc==0]-mean[cc==0])
    meanC = as.vector(meanC)
    varC = Variance[cc==1,cc==1] - Variance[cc==1,cc==0]%*%invObs%*%Variance[cc==0,cc==1]
    if (sum(cc) < 10){
      logden2 = log(pmvnorm(lower[cc==1],upper[cc==1],meanC,NULL,varC)[1])
    } else {
      logden2 = (pmvn(lower[cc==1],upper[cc==1],meanC,varC,uselog2=TRUE)[[1]])/log2(exp(1))
    }
    logver = dmvnorm(c(y[cc==0]),c(mean[cc==0]),as.matrix(Variance[cc==0,cc==0]),TRUE,TRUE) + logden2
  }
  results = list(loglik=logver, AIC=-2*logver+2*(length(beta)+3), BIC=-2*logver+(length(beta)+3)*log(length(y)))
  return (results)

}

# Convergence plot
# -----------------------------------------------------------------------------
plot.convergence = function(model){

  Theta = model$Theta
  X = model$X
  q = length(model$beta)
  Iter = nrow(Theta) - 1

  myplot = vector("list",q+3)
  if (all(X[,1]==1)){ namesE <- c(seq(0,(q-1)),0,0,0) } else { namesE <- c(seq(1,q),0,0,0) }
  Theta1 = rbind(namesE,Theta)
  listabeta = as.list(data.frame(Theta1[,1:q]))
  myplot[1:q] = lapply(listabeta, function(.x) ggplot(data.frame(.x[-1]),aes(x=seq(0,Iter),y=.x[-1])) +
                          geom_line(color="dark blue") + labs(x="Iteration", y=bquote(hat(beta)[.(.x[1])])))
  rm(Theta1)
  myplot[[q+1]] = ggplot(data.frame(Theta),aes(x=seq(0,Iter),y=Theta[,q+1])) + geom_line(color="dark blue") + labs(x="Iteration", y=bquote(hat(sigma)^2))
  myplot[[q+2]] = ggplot(data.frame(Theta),aes(x=seq(0,Iter),y=Theta[,q+2])) + geom_line(color="dark blue") + labs(x="Iteration", y=bquote(hat(phi)))
  myplot[[q+3]] = ggplot(data.frame(Theta),aes(x=seq(0,Iter),y=Theta[,q+3])) + geom_line(color="dark blue") + labs(x="Iteration", y=bquote(hat(tau)^2))
  grid.arrange(grobs=myplot,ncol=3)

}

# Prediction in new locations
# -----------------------------------------------------------------------------
predict.new = function(model,x.new,coord.new){

  mediaO = model$X%*%model$beta
  mediaP = x.new%*%model$beta
  coordAll = rbind(model$coord,coord.new)
  pInd = c(rep(0,length(mediaO)),rep(1,length(mediaP)))
  distM = crossdist(coordAll)
  Variance = varianceMat(model$phi,model$tau2,model$sigma2,model$kappa,distM,model$type)$Sigma
  invObs = solve(Variance[pInd==0,pInd==0])
  yPred = mediaP + Variance[pInd==1,pInd==0]%*%invObs%*%(model$EY - mediaO)
  varPred = Variance[pInd==1,pInd==1] - Variance[pInd==1,pInd==0]%*%invObs%*%Variance[pInd==0,pInd==1]
  return (data.frame(xcoord=coord.new[,1],ycoord=coord.new[,2],predValues=yPred, sdPred=matrix(sqrt(diag(varPred)))))

}

# Generating spatial censored data
# -----------------------------------------------------------------------------
random.cens = function(beta,sigma2,phi,tau,x,coords,type.cens,pcens,npred,type.S,kappa){

  x = as.matrix(x)
  n = nrow(as.matrix(x))
  distM = crossdist(coords)
  varianceC = varianceMat(phi,tau,sigma2,kappa,distM,type.S)$Sigma
  mediaC = x%*%beta
  y = t(rmvnorm(1,mediaC,varianceC))

  # Censored data
  y1 = y[1:(n-npred),]
  if (type.cens=="left"){
    cutoff = quantile(y1,pcens)
    cc = (y1 <= cutoff) + 0
    y1[cc==1] = cutoff
    LI = LS = y1
    LI[cc==1] = -Inf
  } else {
    if (type.cens=="right"){
      cutoff = quantile(y1,1-pcens)
      cc = (y1 >= cutoff) + 0
      y1[cc==1] = cutoff
      LI = LS = y1
      LS[cc==1] = Inf
    }
  } # End if
  TrainingData = data.frame(y1,cc,LI,LS,coords[1:(n-npred),],x[1:(n-npred),])
  names(TrainingData) = c("yobs","cens","LI","LS","xcoord","ycoord",paste("X",1:ncol(x),sep=""))

  if (npred > 0){
    y2 = y[(n-npred+1):n,1]
    TestData = data.frame(y2,coords[(n-npred+1):n,],x[(n-npred+1):n,])
    names(TestData) = c("yobs","xcoord","ycoord",paste("X",1:ncol(x),sep=""))
    return(list(TrainingData=TrainingData,TestData=TestData))
  } else {
    return(TrainingData)
  }

}


# Effective range for spatial correlation
#------------------------------------------------------------------------------
matern.eq <- function(x, cor, phi, kappa){
  mat.eq <- 1/(2^(kappa-1)*gamma(kappa))*(x/phi)^(kappa)*besselK(x/phi,kappa) - cor
  return (mat.eq)
}

Effective.range <- function(cor, phi, kappa, Sp.model) {

  if (Sp.model=="exponential"){
    Eff.range <- -phi*log(cor)
  }

  if (Sp.model=="gaussian"){
    Eff.range <- sqrt(-phi^2*log(cor))
  }

  if (Sp.model=="matern"){
    Eff.range <- uniroot(matern.eq,lower=0.0001,upper=(10000*phi),cor=cor,phi=phi,kappa=kappa)$root
  }

  if (Sp.model=="pow.exp"){
    Eff.range <- (-phi^kappa*log(cor))^(1/kappa)
  }

  Eff.range = as.numeric(Eff.range)
  return (Eff.range)
}


R/rCensspatial_USER.R
#' Censored Spatial Data Simulation
#'
#' This function simulates censored spatial data with a linear structure for an established censoring rate.
#'
#' @param beta linear regression parameters.
#' @param sigma2 partial sill parameter.
#' @param phi spatial scaling parameter.
#' @param nugget nugget effect parameter.
#' @param x design matrix.
#' @param coords 2D spatial coordinates.
#' @param cens '\code{left}' or '\code{right}' censoring. By default \code{='left'}.
#' @param pcens desired censoring rate. By default \code{=0.10}.
#' @param npred number of simulated data used for cross-validation (Prediction). By default \code{=0}.
#' @param cov.model type of spatial correlation function: '\code{exponential}', '\code{gaussian}',
#' '\code{matern}', and '\code{pow.exp}' for exponential, gaussian, matern, and power exponential, respectively.
#' @param kappa parameter for all spatial correlation functions. For exponential and gaussian
#' \eqn{\kappa=0}, for power exponential \eqn{0 < \kappa <= 2}, and for matern correlation function
#' \eqn{\kappa > 0}.
#'
#' @return If \code{npred > 0}, it returns a list with two datasets: \code{TrainingData} and \code{TestData};
#' otherwise, it returns a data frame with the simulated data.
#'
#' \code{TrainingData}
#' \item{yobs}{generated response vector.}
#' \item{cens}{censoring indicator.}
#' \item{LI}{lower censoring bound.}
#' \item{LS}{upper censoring bound.}
#' \item{xcoord}{x coordinates.}
#' \item{ycoord}{y coordinates.}
#' \item{X}{design matrix.}
#'
#' \code{TestData}
#' \item{yobs}{generated response vector.}
#' \item{xcoord}{x coordinates.}
#' \item{ycoord}{y coordinates.}
#' \item{X}{design matrix.}
#'
#' @author Katherine L. Valeriano, Alejandro Ordonez, Christian E. Galarza and Larissa A. Matos.
#'
#' @examples
#' n = 100
#' set.seed(1000)
#' coords = round(matrix(runif(2*n,0,15),n,2),5)
#' x = cbind(1, rnorm(n))
#' data = rCensSp(c(5,2),2,4,0.70,x,coords,"left",0.10,10,"gaussian",0)
#' data$TrainingData
#' data$TestData

rCensSp = function(beta,sigma2,phi,nugget,x,coords,cens="left",pcens=0.10,npred=0,
                   cov.model="exponential",kappa=0){

  beta = c(beta)
  if (!is.numeric(beta)) stop("beta must be a numeric vector.")
  if (!all(is.finite(beta))) stop("beta must contain only finite values.")

  if (length(c(sigma2))>1 | !is.numeric(sigma2)) stop("sigma2 must be a non-negative number.")
  if (sigma2 <= 0) stop("sigma2 must be non-negative.")

  if (length(c(phi))>1 | !is.numeric(phi)) stop("phi must be a non-negative number.")
  if (phi <= 0) stop("phi must be non-negtaive.")

  if (length(c(nugget))>1 | !is.numeric(nugget)) stop("nugget must be a non-negative number.")
  if (nugget <= 0) stop("nugget must be non-negative.")

  x = as.matrix(x)
  if (!all(c(is.finite(x)))) stop("x must contain only finite values.")
  if (ncol(x) != length(c(beta))) stop("Non-conformable dimensions between x and beta.")

  coords = as.matrix(coords)
  if (!all(c(is.finite(coords)))) stop ("coords must contain only finite values.")
  if (ncol(coords) != 2) stop("coords must contain 2 columns.")
  if (nrow(coords) != nrow(x)) stop("Non-conformable dimensions between coords and x.")

  if (is.null(cens)) stop("cens must be specified.")
  if (cens!="left" & cens!="right") stop("cens should be one of left or right.")

  if (!is.numeric(pcens) | length(c(pcens))>1) stop ("pcens must be a real number in (0,1).")
  if (pcens<=0 | pcens>=1) stop("pcens must be a real number in (0,1).")

  if (!is.numeric(npred) | length(c(npred))>1) stop ("npred must be a positive integer.")
  if (npred<0 | npred>=nrow(x) | npred%%1!=0) stop("npred must be a positive integer.")

  if (is.null(cov.model)) stop("cov.model must be specified.")
  if (cov.model!="matern" & cov.model!="gaussian" & cov.model!="pow.exp" & cov.model!="exponential"){
    stop("cov.model should be one of matern, gaussian, pow.exp or exponential.")}

  if (cov.model!="exponential" & cov.model!="gaussian"){
    if (length(c(kappa))>1 | !is.numeric(kappa)) stop("kappa must be specified.")
    if (cov.model=="pow.exp" & (kappa>2| kappa<=0)) stop("kappa must be a real in (0,2].")
    if (cov.model=="matern" & kappa<=0) stop("kappa must be a real number in (0,Inf).")
    if (cov.model=="matern" & is.infinite(kappa)) stop("kappa must be a real number in (0,Inf).")
  }

  #---------------------------------------------------------------------#
  #                                Outputs                              #
  #---------------------------------------------------------------------#

  out.gen = random.cens(beta,sigma2,phi,nugget,x,coords,cens,pcens,npred,cov.model,kappa)
  return(out.gen)

}


R/EstSAEMspatial_USER.R
#' Censored Spatial Model Estimation via SAEM Algorithm
#'
#' This function returns the maximum likelihood (ML) estimates of the unknown parameters in Gaussian spatial models
#' with censored/missing responses via the SAEM algorithm. It supports left, right,
#' interval, or missing values in the dependent variable. It also computes the observed information
#' matrix using the method developed by \insertCite{louis1982finding;textual}{RcppCensSpatial}.
#'
#' @param y vector of responses.
#' @param x design matrix.
#' @param cens vector of censoring indicators. For each observation: \code{1} if censored/missing
#' and \code{0} otherwise.
#' @param LI lower limit of detection. For each observation: if non-censored \code{=y},
#' if left-censored/missing \code{=-Inf}, or \code{=LOD} if right/interval-censored.
#' @param LS upper limit of detection. For each observation: if non-censored \code{=y},
#' if right-censored/missing \code{=Inf}, or \code{=LOD} if left/interval-censored.
#' @param coords 2D spatial coordinates.
#' @param init.phi initial value for the spatial scaling parameter.
#' @param init.nugget initial value for the nugget effect parameter.
#' @param type type of spatial correlation function: '\code{exponential}', '\code{gaussian}',
#' '\code{matern}', and '\code{pow.exp}' for exponential, gaussian, matern, and power exponential, respectively.
#' @param kappa parameter for all spatial correlation functions. See \code{\link{CovMat}}.
#' @param lower,upper vectors of lower and upper bounds for the optimization method. If unspecified, the default is
#' \code{c(0.01,0.01)} for lower and \code{c(30,30)} for upper.
#' @param MaxIter maximum number of iterations of the SAEM algorithm. By default \code{=300}.
#' @param M number of Monte Carlo samples for stochastic approximation. By default \code{=20}.
#' @param pc percentage of iterations of the SAEM algorithm with no-memory. By default \code{=0.25}.
#' @param error maximum convergence error. By default \code{=1e-5}.
#' @param show.SE \code{TRUE} or \code{FALSE}. It indicates if the standard errors should be estimated. By default \code{=TRUE}.
#'
#' @details The spatial Gaussian model is given by
#'
#' \eqn{Y = X\beta + \xi},
#'
#' where \eqn{Y} is the \eqn{n x 1} vector of response, \eqn{X} is the \eqn{n x q} design matrix,
#' \eqn{\beta} is the \eqn{q x 1} vector of regression coefficients to be estimated, and \eqn{\xi}
#' is the error term which is normally distributed with zero-mean and covariance matrix
#' \eqn{\Sigma=\sigma^2 R(\phi) + \tau^2 I_n}. We assume that \eqn{\Sigma} is non-singular and
#' \eqn{X} has full rank \insertCite{diggle2007springer}{RcppCensSpatial}.
#'
#' The estimation process was performed via the SAEM \insertCite{delyon1999convergence}{RcppCensSpatial} algorithm.
#' The spatial SAEM algorithm was previously proposed by \insertCite{lachos2017influence;textual}{RcppCensSpatial}
#' and \insertCite{ordonez2018geostatistical;textual}{RcppCensSpatial} and is available in package \code{CensSpatial}.
#' The difference between this package to \code{CensSpatial} is that the random observations are sampled
#' through the slice sampling algorithm available in package \code{relliptical} and the optimization procedure
#' by the \code{roptim} package.
#'
#' This model is also a particular case of the Spatio-temporal model defined by \insertCite{valeriano2021likelihood;textual}{RcppCensSpatial},
#' when the number of temporal observations is equal to one. The computing codes of the Spatio-temporal
#' SAEM algorithm are available in the package \code{StempCens}.
#'
#' @note
#' The SAEM final estimates correspond to the estimates obtained at the last iteration of the
#' algorithm.
#'
#' To fit a regression model for non-censored data, just set \code{cens} as a vector of zeros.
#'
#' Functions \code{print}, \code{summary}, and \code{plot} work for objects of class \code{sclm}.
#'
#' @return The function returns an object of class \code{sclm} which is a list given by:
#'
#' \item{Theta}{estimated parameters in all iterations, \eqn{\theta = (\beta, \sigma^2, \phi, \tau^2)}.}
#' \item{theta}{final estimation of \eqn{\theta = (\beta, \sigma^2, \phi, \tau^2)}.}
#' \item{beta}{estimated \eqn{\beta}.}
#' \item{sigma2}{estimated \eqn{\sigma^2}.}
#' \item{phi}{estimated \eqn{\phi}.}
#' \item{tau2}{estimated \eqn{\tau^2}.}
#' \item{EY}{stochastic approximation of the first moment for the truncated normal distribution.}
#' \item{EYY}{stochastic approximation of the second moment for the truncated normal distribution.}
#' \item{SE}{vector of standard errors of \eqn{\theta = (\beta, \sigma^2, \phi, \tau^2)}.}
#' \item{InfMat}{observed information matrix.}
#' \item{loglik}{log-likelihood for the SAEM method.}
#' \item{AIC}{Akaike information criterion.}
#' \item{BIC}{Bayesian information criterion.}
#' \item{Iterations}{number of iterations needed to converge.}
#' \item{ptime}{processing time.}
#' \item{range}{the effective range.}
#'
#' @author Katherine L. Valeriano, Alejandro Ordonez, Christian E. Galarza and Larissa A. Matos.
#'
#' @seealso \code{\link{EM.sclm}}, \code{\link{MCEM.sclm}}, \code{\link{predict.sclm}}
#'
#' @examples
#' # Simulated example: 10% of right-censored observations
#' n = 50   # Test with another values for n
#' set.seed(1000)
#' coords = round(matrix(runif(2*n,0,15),n,2),5)
#' x = cbind(rbinom(n,1,0.50), rnorm(n), rnorm(n))
#' data = rCensSp(c(1,4,-2),2,3,0.50,x,coords,"right",0.10,0,"matern",2)
#'
#' fit = SAEM.sclm(y=data$yobs, x=data[,7:9], cens=data$cens, LI=data$LI,
#'              LS=data$LS, coords=data[,5:6], init.phi=2, init.nugget=1,
#'              type="matern", kappa=2, MaxIter = 20, error=1e-4)
#' summary(fit)
#'
#' \donttest{
#' # Simulated example: censored and missing observations
#' n = 200
#' set.seed(123)
#' coords = round(matrix(runif(2*n,0,20),n,2),5)
#' x = cbind(1, rnorm(n), rexp(n))
#' data = rCensSp(c(1,4,-1),2,4,0.50,x,coords,"left",0.10,0,"exponential",0)
#' data$yobs[c(10,20)] = NA;   data$cens[c(10,20)] = 1
#' data$LI[c(10,20)] = -Inf;   data$LS[c(10,20)] = Inf
#'
#' fit2 = SAEM.sclm(y=data$yobs, x=data[,7:9], cens=data$cens, LI=data$LI,
#'               LS=data$LS, coords=data[,5:6], init.phi=2, init.nugget=1,
#'               type="exponential", MaxIter = 300, error=1e-4)
#' fit2$theta  # Estimates
#' fit2$SE     # Standard error
#' fit2$InfMat # Information matrix
#' plot(fit2)}
#' @references
#' \insertAllCited

SAEM.sclm = function(y,x,cens,LI,LS,coords,init.phi,init.nugget,type="exponential",kappa=0,
                          lower=c(0.01,0.01),upper=c(30,30),MaxIter=300,M=20,pc=0.25,
                          error=1e-5,show.SE=TRUE){

  #---------------------------------------------------------------------#
  #                              Validations                            #
  #---------------------------------------------------------------------#
  y = as.matrix(y);     n = nrow(y)
  x = as.matrix(x);     cens = as.matrix(cens)
  LI = as.matrix(LI);   LS = as.matrix(LS)
  coords = as.matrix(coords)

  if (!is.numeric(y)) stop("y must be a numeric vector.")
  if (ncol(y) > 1) stop("y must have just one column.")

  if (!all(c(is.finite(x)))) stop("x must contain only finite values.")
  if (nrow(x) != n) stop("Non-conformable dimensions between x and y.")

  if (nrow(cens) != n) stop("Non-conformable dimensions between cens and y.")
  if (ncol(cens) > 1) stop("cens must have just one column.")
  if (!all(c(cens)==0 | c(cens)==1)) stop("cens must contain only 0 or 1.")
  if (!all(cens[is.na(y)]==1)) stop("cens must be equal 1 when y is NA.")

  if (nrow(LI)!=n | nrow(LS)!=n) stop("Non-conformable dimensions on LI or LS.")
  if (ncol(LI)>1 | ncol(LS)>1) stop("LI and LS must have just one column.")
  if (any(is.na(LI)) | !is.numeric(LI)) stop("LI is not specified or contains NA.")
  if (any(is.na(LS)) | !is.numeric(LS)) stop("LS is not specified or contains NA.")
  if (any(LI[cens==1]>=LS[cens==1])) stop("LI must be lower or equal than LS.")
  if (sum(is.na(y)) > 0){ LI[is.na(y)] = -Inf;  LS[is.na(y)] = Inf }

  if (!all(c(is.finite(coords)))) stop("coords must contain only finite values.")
  if (nrow(coords)!=n | ncol(coords)!=2) stop("Non-conformable dimensions between coords and y.")

  if (length(c(init.phi))>1 | !is.numeric(init.phi)) stop("Initial value for phi must be provided.")
  if (init.phi <= 0) stop("init.phi must be non-negative.")

  if (length(c(init.nugget))>1 | !is.numeric(init.nugget)) stop("Initial value for nugget effect must be provided.")
  if (init.nugget <= 0) stop("init.nugget must be non-negative.")

  if (is.null(type)) stop("type must be specified.")
  if (type!="matern" & type!="gaussian" & type!="pow.exp" & type!="exponential"){
    stop("type should be one of matern, gaussian, pow.exp or exponential.")}

  if (type!="exponential" & type!="gaussian"){
    if (length(c(kappa))>1 | !is.numeric(kappa)) stop("kappa must be specified.")
    if (type=="pow.exp" & (kappa>2| kappa<=0)) stop("kappa must be a real in (0,2].")
    if (type=="matern" & kappa<=0) stop("kappa must be a real number in (0,Inf).")
    if (type=="matern" & is.infinite(kappa)) stop("kappa must be a real number in (0,Inf).")
  }

  if (length(c(lower))!=2 | length(c(upper))!=2) stop("lower and upper must be vectors of length 2.")
  if (any(is.na(lower)) | !is.numeric(lower)) stop("lower is not specified or contains NA.")
  if (any(is.na(upper)) | !is.numeric(upper)) stop("upper is not specified or contains NA.")
  if (any(lower>=upper)) stop("lower must be smaller than upper.")
  if (any(lower<=0)) stop("Values in lower must be non-negative.")
  if (!all(is.finite(upper))) stop("upper must contain only finite values.")

  if (length(c(MaxIter))>1 | !is.numeric(MaxIter)) stop("MaxIter must be a positive integer.")
  if (MaxIter<=0 | MaxIter%%1!=0) stop("MaxIter must be a positive integer.")

  if (length(c(M))>1 | !is.numeric(M)) stop("M must be a positive integer.")
  if (M<=0 | M%%1!=0) stop("M must be a positive integer.")

  if (length(c(pc))>1 | !is.numeric(pc)) stop("pc must belong to the interval [0,1].")
  if (pc<0 | pc>1) stop("pc must belong to the interval [0,1].")

  if (length(c(error))>1 | !is.numeric(error)) stop("error must be specified.")
  if (error<=0 | error>1) stop("error must belong to the interval (0,1].")

  if (!is.logical(show.SE)) stop("show.SE must be logical (TRUE/FALSE).")

  #---------------------------------------------------------------------#
  #                                Outputs                              #
  #---------------------------------------------------------------------#

  out.Sp = SAEM_Spatial(y,x,cens,LI,LS,coords,init.phi,init.nugget,type,kappa,
                                         lower,upper,MaxIter,M,pc,error,show.SE)

  out.Sp$call = match.call()
  out.Sp$range  = Effective.range(0.05,out.Sp$phi,kappa,type)
  out.Sp$show.SE = show.SE
  out.Sp$cens = cens
  out.Sp$MaxIter = MaxIter

  class(out.Sp) <- "sclm"
  return(out.Sp)

}
